### PR TITLE
Add feature to skip certain interfaces when detecting HW type

### LINKED ISSUE #31: Metals with any interface (e.g. virbr0) with MAC prefix 52:54:00 are incorrectly recognized as kvm type
I have a metal machine with the following MAC addresses:
```
$ juju show-machine 0
...
    network-interfaces:
      bond0:
        ip-addresses:
        - 10.56.150.113
        mac-address: ac:1f:6b:32:12:dc
        gateway: 10.56.150.1
        space: oam-space
        is-up: true
      br-bond0:
        ip-addresses:
        - 10.56.150.113
        mac-address: ac:1f:6b:32:12:dc
        gateway: 10.56.150.1
        space: oam-space
        is-up: true
      virbr0:
        ip-addresses:
        - 192.168.122.1
        mac-address: 52:54:00:31:a0:dd
        space: undefined
        is-up: true

```
Because there's a virbr0 bridge with 52:54:00 MAC address prefix it gets incorrectly recognized as a KVM. There's no option to override this as we can currently do with KVMs incorrectly recognized as metals.
This mainly affects MAAS nodes that often host KVMs and hence the virbr0 bridge exists on them, but there were occurrences of regular OpenStack compute nodes incorrectly recognized as KVMs due to the same issue (virbr0 with KVM-specific MAC address).

### PR DESCRIPTION
**Pull request by mkalcok**
Adds ability to specify to specify list of regular expressions. Interfaces whose names match any regexp in this list will not be considered when determining hardware type of the machine.

Closes #31

### CODE REVIEW

**Review by rgildein — APPROVED** [review]
LGTM

**Review by esunar — APPROVED** [review]

File: prometheus_juju_exporter/config_default.yaml [hunk 1]
```diff
@@ -13,9 +13,10 @@ exporter:
   collect_interval: 15
 
 detection: # parameters affecting the detection algorithm
-  virt_macs: 
+  skip_interfaces: []  # regexp list of interfaces that should be skipped during hw typ detection
```
Comment (agileshaw) [inline prometheus_juju_exporter/config_default.yaml:16]: I think we should provide a default list of interfaces here, similar to the `virt_macs`. We cannot assume the snap will always be used with the charm, and these interfaces should be ignored in most (if not all) cases.

**Review by agileshaw — CHANGES_REQUESTED** [review]
Overall LGTM. Just one suggestion inline

File: prometheus_juju_exporter/config_default.yaml (hunk 1 above)
Comment (agileshaw) [reply prometheus_juju_exporter/config_default.yaml:16]: After discussion, we agreed to keep the `skip_interface` config option empty in the snap config and let charm take charge of setting it.

For the `virt_macs` option, we will open another PR to remove its default value.

**Review by agileshaw — APPROVED** [review]
LGTM

File: prometheus_juju_exporter/collector.py [hunk 2]
```diff
@@ -125,20 +126,36 @@ def _get_labels_to_remove(self, gauge_name: str) -> None:
...
         """Detect machine type based on its MAC address.
 
         :param dict machine: status information for a machine
         :return MachineType: MachineType class object indicating the machine type
         """
-        interfaces = machine["network-interfaces"].values()
+        self.logger.debug("Determining type of machine: %s", machine_id)
+        machine_type = MachineType.METAL
+        skip_interfaces = self.config["detection"]["skip_interfaces"].as_str_seq()
+        interface_filter = re.compile(r"(?=(" + "|".join(skip_interfaces) + r"))")
```
Comment (aieri) [inline prometheus_juju_exporter/collector.py:138]: why is the lookahead needed?

File: tests/unit/test_collector.py [hunk 3]
```diff
@@ -189,10 +191,43 @@ def test_get_machine_type(self, collector_daemon, mac_address, expect_machine_ty
...
         }
-        machine_type = statsd._get_machine_type(machine)
+        machine_type = statsd._get_machine_type(machine, "dummy-0")
 
         assert machine_type.value == expect_machine_type
 
+    def test_get_machine_type_interface_skip(self, collector_daemon):
+        """Test that blacklisted interfaces are not used to detect machine type."""
+        statsd = collector_daemon()
+        skip_interfaces = ["virb*", "tap-*"]
```
Comment (aieri) [inline tests/unit/test_collector.py:201]: these will match anyway, but you probably meant `virb.*` and `tap-.*`

File: tests/unit/test_collector.py [hunk 4]
```diff
@@ -189,10 +191,43 @@ def test_get_machine_type(self, collector_daemon, mac_address, expect_machine_ty
...
+                },
+                "tap-0": {
+                    "mac-address": tap_prefix + "00:00:01",
+                },
+            }
+        }
+
+        statsd.config["detection"]["virt_macs"].set([kvm_prefix, tap_prefix])
+        # Assert that without skipping interfaces, machine is detected as virtual
+        assert statsd._get_machine_type(machine, "dummy-0") == MachineType.KVM
```
Comment (aieri) [inline tests/unit/test_collector.py:225]: it's better to have a single assert per test (so it's clearer what broke when one test fails). I would suggest splitting this test into two (perhaps parametrizing if convenient)

**Review by aieri — CHANGES_REQUESTED** [review]
Generally looking good, I've only requested a change in the testing code and added a couple of small comments.

File: tests/unit/test_collector.py (hunk 3 above)
Comment (mkalcok) [reply tests/unit/test_collector.py:201]: Come to think of it, `["^virbr", "^tap-"]` are probably good enough, right?

Comment (aieri) [reply tests/unit/test_collector.py:201]: they will all match, even just `["virbr", "tap"]`; it depends on how specific you want them to be. It's just that `virb*` and `tap-*` look like typos and not what you meant

File: prometheus_juju_exporter/collector.py (hunk 2 above)
Comment (mkalcok) [reply prometheus_juju_exporter/collector.py:138]: Truth be told, regexps are always confusing to me, so I googled around and this looked like it did the job. But you are right. I tested it now and 
```python
interface_filter = re.compile(r"|".join(skip_interfaces))
```
should do the trick, correct?

Comment (aieri) [reply prometheus_juju_exporter/collector.py]: I would expect it to, yes. The positive lookahead is useful for expressing "match this $THING if it is followed by this $OTHERTHING"; in your case you simply want "match one of these patterns"

File: tests/unit/test_collector.py [hunk 5]
```diff
@@ -195,10 +195,23 @@ def test_get_machine_type(self, collector_daemon, mac_address, expect_machine_ty
 
         assert machine_type.value == expect_machine_type
 
-    def test_get_machine_type_interface_skip(self, collector_daemon):
-        """Test that blacklisted interfaces are not used to detect machine type."""
+    @pytest.mark.parametrize(
+        "skip_interfaces, expected_type",
+        [
+            ([r"^virbr\d*", r"^tap-"], MachineType.METAL),
```
Comment (aieri) [inline tests/unit/test_collector.py:201]: interesting, does the `r` prefix not cause problems? The config option should be a plain string afaiu

**Review by aieri — COMMENTED** [review]
LGTM, just one small non-blocking comment

File: tests/unit/test_collector.py (hunk 5 above)
Comment (mkalcok) [reply tests/unit/test_collector.py:201]: It throws warnings without the `r` complaining about invalid escape sequence. but tests pass either way.

**Review by aieri — APPROVED** [review]

**Pull request merged**